File: cocos/cocos2d-x/cocos/platform/android/java/src/org/cocos2dx/lib/CanvasRenderingContext2DImpl.java
/****************************************************************************
 * Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.
 *
 * http://www.cocos.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ****************************************************************************/

package eggy.cocos2dx.lib;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Typeface;
import android.os.Build;
import android.text.TextPaint;
import android.util.Log;

import java.lang.ref.WeakReference;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;

public class CanvasRenderingContext2DImpl {

    private static final String TAG = "CanvasContext2D";

    private static final int TEXT_ALIGN_LEFT = 0;
    private static final int TEXT_ALIGN_CENTER = 1;
    private static final int TEXT_ALIGN_RIGHT = 2;

    private static final int TEXT_BASELINE_TOP = 0;
    private static final int TEXT_BASELINE_MIDDLE = 1;
    private static final int TEXT_BASELINE_BOTTOM = 2;

    private static WeakReference<Context> sContext;
    private TextPaint mTextPaint;
    private Paint mLinePaint;
    private Path mLinePath;
    private final Canvas mCanvas = new Canvas();
    private Bitmap mBitmap;
    private int mTextAlign = TEXT_ALIGN_LEFT;
    private int mTextBaseline = TEXT_BASELINE_BOTTOM;
    private int mFillStyleR = 0;
    private int mFillStyleG = 0;
    private int mFillStyleB = 0;
    private int mFillStyleA = 255;

    private int mStrokeStyleR = 0;
    private int mStrokeStyleG = 0;
    private int mStrokeStyleB = 0;
    private int mStrokeStyleA = 255;

    private String mFontName = "Arial";
    private float mFontSize = 40.0f;
    private float mLineWidth = 0.0f;
    private static final float _sApproximatingOblique = -0.25f;// please check paint api documentation
    private boolean mIsBoldFont = false;
    private boolean mIsItalicFont = false;
    private boolean mIsObliqueFont = false;
    private boolean mIsSmallCapsFontVariant = false;
    private String mLineCap = "butt";
    private String mLineJoin = "miter";

    private class Size {
        Size(float w, float h) {
            this.width = w;
            this.height = h;
        }

        Size() {
            this.width = 0;
            this.height = 0;
        }

        public float width;
        public float height;
    }

    private class Point {
        Point(float x, float y) {
            this.x = x;
            this.y = y;
        }

        Point() {
            this.x = this.y = 0.0f;
        }

        Point(Point pt) {
            this.x = pt.x;
            this.y = pt.y;
        }

        void set(float x, float y) {
            this.x = x;
            this.y = y;
        }

        public float x;
        public float y;
    }

    static void init(Context context) {
        sContext = new WeakReference<>(context);
    }

    static void destroy() {
        sContext = null;
    }

    private static final HashMap<String, Typeface> sTypefaceCache = new HashMap<>();

    private static void loadTypeface(String familyName, String url) {
        if (!sTypefaceCache.containsKey(familyName)) {
            try {
                Typeface typeface = null;
                if (url.startsWith("/")) {
                    typeface = Typeface.createFromFile(url);
                } else if (sContext.get() != null) {
                    final String prefix = "@assets/";
                    if (url.startsWith(prefix)) {
                        url = url.substring(prefix.length());
                    }
                    typeface = Typeface.createFromAsset(sContext.get().getAssets(), url);
                }

                if (typeface != null) {
                    sTypefaceCache.put(familyName, typeface);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static void clearTypefaceCache() {
        sTypefaceCache.clear();
    }

    private static TextPaint newPaint(String fontName, int fontSize, boolean enableBold, boolean enableItalic,
            boolean obliqueFont, boolean smallCapsFontVariant) {
        TextPaint paint = new TextPaint();
        paint.setTextSize(fontSize);
        paint.setAntiAlias(true);
        paint.setSubpixelText(true);

        String key = fontName;
        if (enableBold) {
            key += "-Bold";
            paint.setFakeBoldText(true);
        }
        if (enableItalic) {
            key += "-Italic";
        }

        Typeface typeFace;
        if (sTypefaceCache.containsKey(key)) {
            typeFace = sTypefaceCache.get(key);
        } else {
            int style = Typeface.NORMAL;
            if (enableBold && enableItalic) {
                style = Typeface.BOLD_ITALIC;
            } else if (enableBold) {
                style = Typeface.BOLD;
            } else if (enableItalic) {
                style = Typeface.ITALIC;
            }
            typeFace = Typeface.create(fontName, style);
        }
        paint.setTypeface(typeFace);
        if (obliqueFont) {
            paint.setTextSkewX(_sApproximatingOblique);
        }
        if (smallCapsFontVariant && Build.VERSION.SDK_INT >= 21) {
            Cocos2dxReflectionHelper.<Void>invokeInstanceMethod(paint,
                    "setFontFeatureSettings",
                    new Class[] { String.class },
                    new Object[] { "smcp" });
        }
        return paint;
    }

    private CanvasRenderingContext2DImpl() {
    }

    private void recreateBuffer(float w, float h) {
        if (mBitmap != null) {
            mBitmap.recycle();
        }
        mBitmap = Bitmap.createBitmap((int) Math.ceil(w), (int) Math.ceil(h), Bitmap.Config.ARGB_8888);
        mCanvas.setBitmap(mBitmap);
    }

    private void beginPath() {
        if (mLinePath == null) {
            mLinePath = new Path();
        }
        mLinePath.reset();
    }

    private void closePath() {
        mLinePath.close();
    }

    private void moveTo(float x, float y) {
        mLinePath.moveTo(x, y);
    }

    private void lineTo(float x, float y) {
        mLinePath.lineTo(x, y);
    }

    private void stroke() {
        if (mLinePaint == null) {
            mLinePaint = new Paint();
            mLinePaint.setAntiAlias(true);
        }

        if (mLinePath == null) {
            mLinePath = new Path();
        }

        mLinePaint.setARGB(mStrokeStyleA, mStrokeStyleR, mStrokeStyleG, mStrokeStyleB);
        mLinePaint.setStyle(Paint.Style.STROKE);
        mLinePaint.setStrokeWidth(mLineWidth);
        this.setStrokeCap(mLinePaint);
        this.setStrokeJoin(mLinePaint);
        mCanvas.drawPath(mLinePath, mLinePaint);
    }

    private void setStrokeCap(Paint paint) {
        switch (mLineCap) {
            case "butt":
                paint.setStrokeCap(Paint.Cap.BUTT);
                break;
            case "round":
                paint.setStrokeCap(Paint.Cap.ROUND);
                break;
            case "square":
                paint.setStrokeCap(Paint.Cap.SQUARE);
                break;
        }
    }

    private void setStrokeJoin(Paint paint) {
        switch (mLineJoin) {
            case "bevel":
                paint.setStrokeJoin(Paint.Join.BEVEL);
                break;
            case "round":
                paint.setStrokeJoin(Paint.Join.ROUND);
                break;
            case "miter":
                paint.setStrokeJoin(Paint.Join.MITER);
                break;
        }
    }

    private void fill() {
        if (mLinePaint == null) {
            mLinePaint = new Paint();
        }

        if (mLinePath == null) {
            mLinePath = new Path();
        }

        mLinePaint.setARGB(mFillStyleA, mFillStyleR, mFillStyleG, mFillStyleB);
        mLinePaint.setStyle(Paint.Style.FILL);
        mCanvas.drawPath(mLinePath, mLinePaint);
        mLinePaint.setStrokeWidth(0);
        this.setStrokeCap(mLinePaint);
        this.setStrokeJoin(mLinePaint);
        mLinePaint.setStyle(Paint.Style.STROKE);
        mCanvas.drawPath(mLinePath, mLinePaint);
        mLinePaint.setStrokeWidth(mLineWidth);
    }

    private void setLineCap(String lineCap) {
        mLineCap = lineCap;
    }

    private void setLineJoin(String lineJoin) {
        mLineJoin = lineJoin;
    }

    private void saveContext() {
        mCanvas.save();
    }

    private void restoreContext() {
        if (mCanvas.getSaveCount() > 1) {
            mCanvas.restore();
        }
    }

    private void rect(float x, float y, float w, float h) {
        beginPath();
        moveTo(x, y);
        lineTo(x, y + h);
        lineTo(x + w, y + h);
        lineTo(x + w, y);
        closePath();
    }

    private void clearRect(float x, float y, float w, float h) {
        int clearSize = (int) (w * h);
        int[] clearColor = new int[clearSize];
        for (int i = 0; i < clearSize; ++i) {
            clearColor[i] = Color.TRANSPARENT;
        }
        mBitmap.setPixels(clearColor, 0, (int) w, (int) x, (int) y, (int) w, (int) h);
    }

    private void createTextPaintIfNeeded() {
        if (mTextPaint == null) {
            mTextPaint = newPaint(mFontName, (int) mFontSize, mIsBoldFont, mIsItalicFont, mIsObliqueFont,
                    mIsSmallCapsFontVariant);
        }
    }

    private void fillRect(float x, float y, float w, float h) {
        int pixelValue = (mFillStyleA & 0xff) << 24 | (mFillStyleR & 0xff) << 16 | (mFillStyleG & 0xff) << 8
                | (mFillStyleB & 0xff);
        int fillSize = (int) (w * h);
        int[] fillColors = new int[fillSize];
        for (int i = 0; i < fillSize; ++i) {
            fillColors[i] = pixelValue;
        }
        mBitmap.setPixels(fillColors, 0, (int) w, (int) x, (int) y, (int) w, (int) h);
    }

    private void scaleX(TextPaint textPaint, String text, float maxWidth) {
        if (maxWidth < Float.MIN_VALUE)
            return;
        float measureWidth = this.measureText(text);
        if ((measureWidth - maxWidth) < Float.MIN_VALUE)
            return;
        float scaleX = maxWidth / measureWidth;
        textPaint.setTextScaleX(scaleX);
    }

    private void fillText(String text, float x, float y, float maxWidth) {
        createTextPaintIfNeeded();
        mTextPaint.setARGB(mFillStyleA, mFillStyleR, mFillStyleG, mFillStyleB);
        mTextPaint.setStyle(Paint.Style.FILL);
        scaleX(mTextPaint, text, maxWidth);
        Point pt = convertDrawPoint(new Point(x, y), text);
        mCanvas.drawText(text, pt.x, pt.y, mTextPaint);
    }

    private void strokeText(String text, float x, float y, float maxWidth) {
        createTextPaintIfNeeded();
        mTextPaint.setARGB(mStrokeStyleA, mStrokeStyleR, mStrokeStyleG, mStrokeStyleB);
        mTextPaint.setStyle(Paint.Style.STROKE);
        mTextPaint.setStrokeWidth(mLineWidth);
        scaleX(mTextPaint, text, maxWidth);
        Point pt = convertDrawPoint(new Point(x, y), text);
        mCanvas.drawText(text, pt.x, pt.y, mTextPaint);
    }

    private float measureText(String text) {
        createTextPaintIfNeeded();
        float ret = mTextPaint.measureText(text);
        return ret;
    }

    private Size measureTextReturnSize(String text) {
        createTextPaintIfNeeded();
        Paint.FontMetrics fm = mTextPaint.getFontMetrics();
        return new Size(measureText(text), fm.descent - fm.ascent);
    }

    private void updateFont(String fontName, float fontSize, boolean bold, boolean italic, boolean oblique,
            boolean smallCaps) {
        mFontName = fontName;
        mFontSize = fontSize;
        mIsBoldFont = bold;
        mIsItalicFont = italic;
        mIsObliqueFont = oblique;
        mIsSmallCapsFontVariant = smallCaps;
        mTextPaint = null; // Reset paint to re-create paint object in createTextPaintIfNeeded
    }

    private void setTextAlign(int align) {
        mTextAlign = align;
    }

    private void setTextBaseline(int baseline) {
        mTextBaseline = baseline;
    }

    private void setFillStyle(float r, float g, float b, float a) {
        mFillStyleR = (int) (r * 255.0f);
        mFillStyleG = (int) (g * 255.0f);
        mFillStyleB = (int) (b * 255.0f);
        mFillStyleA = (int) (a * 255.0f);
    }

    private void setStrokeStyle(float r, float g, float b, float a) {
        mStrokeStyleR = (int) (r * 255.0f);
        mStrokeStyleG = (int) (g * 255.0f);
        mStrokeStyleB = (int) (b * 255.0f);
        mStrokeStyleA = (int) (a * 255.0f);
    }

    private void setLineWidth(float lineWidth) {
        mLineWidth = lineWidth;
    }

    private void _fillImageData(byte[] imageData, float imageWidth, float imageHeight, float offsetX, float offsetY) {
        int fillSize = (int) (imageWidth * imageHeight);
        int[] fillColors = new int[fillSize];
        int r, g, b, a;
        for (int i = 0; i < fillSize; ++i) {
            r = ((int) imageData[4 * i + 0]) & 0xff;
            g = ((int) imageData[4 * i + 1]) & 0xff;
            b = ((int) imageData[4 * i + 2]) & 0xff;
            a = ((int) imageData[4 * i + 3]) & 0xff;
            fillColors[i] = (a & 0xff) << 24 | (r & 0xff) << 16 | (g & 0xff) << 8 | (b & 0xff);
        }
        mBitmap.setPixels(fillColors, 0, (int) imageWidth, (int) offsetX, (int) offsetY, (int) imageWidth,
                (int) imageHeight);
    }

    private Point convertDrawPoint(final Point point, String text) {
        Point ret = new Point(point);
        Size textSize = measureTextReturnSize(text);

        if (mTextAlign == TEXT_ALIGN_CENTER) {
            ret.x -= textSize.width / 2;
        } else if (mTextAlign == TEXT_ALIGN_RIGHT) {
            ret.x -= textSize.width;
        }

        if (mTextBaseline == TEXT_BASELINE_TOP) {
            ret.y += textSize.height;
        } else if (mTextBaseline == TEXT_BASELINE_MIDDLE) {
            ret.y += textSize.height / 2;
        }

        return ret;
    }

    private byte[] getDataRef() {
        if (mBitmap != null) {
            final int len = mBitmap.getWidth() * mBitmap.getHeight() * 4;
            final byte[] pixels = new byte[len];
            final ByteBuffer buf = ByteBuffer.wrap(pixels);
            buf.order(ByteOrder.nativeOrder());
            mBitmap.copyPixelsToBuffer(buf);

            return pixels;
        }

        return null;
    }
}

File: lib-engine/src/main/kotlin/eggy/cocos2dx/lib/CanvasRenderingContext2DImpl.kt
/****************************************************************************
 * Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.
 *
 * http://www.cocos.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package eggy.cocos2dx.lib

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Typeface
import android.os.Build
import android.text.TextPaint
import eggy.manager.CocosAssetsMgr
import java.lang.ref.WeakReference
import java.nio.ByteBuffer
import java.nio.ByteOrder

class CanvasRenderingContext2DImpl private constructor() {
    private var mTextPaint: TextPaint? = null
    private var mLinePaint: Paint? = null
    private var mLinePath: Path? = null
    private val mCanvas = Canvas()
    private var mBitmap: Bitmap? = null
    private var mTextAlign = TEXT_ALIGN_LEFT
    private var mTextBaseline = TEXT_BASELINE_BOTTOM
    private var mFillStyleR = 0
    private var mFillStyleG = 0
    private var mFillStyleB = 0
    private var mFillStyleA = 255
    private var mStrokeStyleR = 0
    private var mStrokeStyleG = 0
    private var mStrokeStyleB = 0
    private var mStrokeStyleA = 255
    private var mFontName = "Arial"
    private var mFontSize = 40.0f
    private var mLineWidth = 0.0f
    private var mIsBoldFont = false
    private var mIsItalicFont = false
    private var mIsObliqueFont = false
    private var mIsSmallCapsFontVariant = false
    private var mLineCap = "butt"
    private var mLineJoin = "miter"

    private inner class Size {
        internal constructor(w: Float, h: Float) {
            width = w
            height = h
        }

        internal constructor() {
            width = 0f
            height = 0f
        }

        var width: Float
        var height: Float
    }

    private inner class Point {
        internal constructor(x: Float, y: Float) {
            this.x = x
            this.y = y
        }

        internal constructor() {
            y = 0.0f
            x = y
        }

        internal constructor(pt: Point) {
            x = pt.x
            y = pt.y
        }

        operator fun set(x: Float, y: Float) {
            this.x = x
            this.y = y
        }

        var x: Float
        var y: Float
    }

    private fun recreateBuffer(w: Float, h: Float) {
        if (mBitmap != null) {
            mBitmap!!.recycle()
        }
        mBitmap = Bitmap.createBitmap(
            Math.ceil(w.toDouble()).toInt(),
            Math.ceil(h.toDouble()).toInt(),
            Bitmap.Config.ARGB_8888
        )
        mCanvas.setBitmap(mBitmap)
    }

    private fun beginPath() {
        if (mLinePath == null) {
            mLinePath = Path()
        }
        mLinePath!!.reset()
    }

    private fun closePath() {
        mLinePath!!.close()
    }

    private fun moveTo(x: Float, y: Float) {
        mLinePath!!.moveTo(x, y)
    }

    private fun lineTo(x: Float, y: Float) {
        mLinePath!!.lineTo(x, y)
    }

    private fun stroke() {
        if (mLinePaint == null) {
            mLinePaint = Paint()
            mLinePaint!!.isAntiAlias = true
        }
        if (mLinePath == null) {
            mLinePath = Path()
        }
        mLinePaint!!.setARGB(mStrokeStyleA, mStrokeStyleR, mStrokeStyleG, mStrokeStyleB)
        mLinePaint!!.style = Paint.Style.STROKE
        mLinePaint!!.strokeWidth = mLineWidth
        setStrokeCap(mLinePaint!!)
        setStrokeJoin(mLinePaint!!)
        mCanvas.drawPath(mLinePath!!, mLinePaint!!)
    }

    private fun setStrokeCap(paint: Paint) {
        when (mLineCap) {
            "butt" -> paint.strokeCap = Paint.Cap.BUTT
            "round" -> paint.strokeCap = Paint.Cap.ROUND
            "square" -> paint.strokeCap = Paint.Cap.SQUARE
        }
    }

    private fun setStrokeJoin(paint: Paint) {
        when (mLineJoin) {
            "bevel" -> paint.strokeJoin = Paint.Join.BEVEL
            "round" -> paint.strokeJoin = Paint.Join.ROUND
            "miter" -> paint.strokeJoin = Paint.Join.MITER
        }
    }

    private fun fill() {
        if (mLinePaint == null) {
            mLinePaint = Paint()
        }
        if (mLinePath == null) {
            mLinePath = Path()
        }
        mLinePaint!!.setARGB(mFillStyleA, mFillStyleR, mFillStyleG, mFillStyleB)
        mLinePaint!!.style = Paint.Style.FILL
        mCanvas.drawPath(mLinePath!!, mLinePaint!!)
        mLinePaint!!.strokeWidth = 0f
        setStrokeCap(mLinePaint!!)
        setStrokeJoin(mLinePaint!!)
        mLinePaint!!.style = Paint.Style.STROKE
        mCanvas.drawPath(mLinePath!!, mLinePaint!!)
        mLinePaint!!.strokeWidth = mLineWidth
    }

    private fun setLineCap(lineCap: String) {
        mLineCap = lineCap
    }

    private fun setLineJoin(lineJoin: String) {
        mLineJoin = lineJoin
    }

    private fun saveContext() {
        mCanvas.save()
    }

    private fun restoreContext() {
        if (mCanvas.saveCount > 1) {
            mCanvas.restore()
        }
    }

    private fun rect(x: Float, y: Float, w: Float, h: Float) {
        beginPath()
        moveTo(x, y)
        lineTo(x, y + h)
        lineTo(x + w, y + h)
        lineTo(x + w, y)
        closePath()
    }

    private fun clearRect(x: Float, y: Float, w: Float, h: Float) {
        val clearSize = (w * h).toInt()
        val clearColor = IntArray(clearSize)
        for (i in 0 until clearSize) {
            clearColor[i] = Color.TRANSPARENT
        }
        mBitmap!!.setPixels(clearColor, 0, w.toInt(), x.toInt(), y.toInt(), w.toInt(), h.toInt())
    }

    private fun createTextPaintIfNeeded() {
        if (mTextPaint == null) {
            mTextPaint = newPaint(
                mFontName,
                mFontSize.toInt(),
                mIsBoldFont,
                mIsItalicFont,
                mIsObliqueFont,
                mIsSmallCapsFontVariant
            )
        }
    }

    private fun fillRect(x: Float, y: Float, w: Float, h: Float) {
        val pixelValue =
            mFillStyleA and 0xff shl 24 or (mFillStyleR and 0xff shl 16) or (mFillStyleG and 0xff shl 8) or (mFillStyleB and 0xff)
        val fillSize = (w * h).toInt()
        val fillColors = IntArray(fillSize)
        for (i in 0 until fillSize) {
            fillColors[i] = pixelValue
        }
        mBitmap!!.setPixels(fillColors, 0, w.toInt(), x.toInt(), y.toInt(), w.toInt(), h.toInt())
    }

    private fun scaleX(textPaint: TextPaint?, text: String, maxWidth: Float) {
        if (maxWidth < Float.MIN_VALUE) return
        val measureWidth = measureText(text)
        if (measureWidth - maxWidth < Float.MIN_VALUE) return
        val scaleX = maxWidth / measureWidth
        textPaint!!.textScaleX = scaleX
    }

    private fun fillText(text: String, x: Float, y: Float, maxWidth: Float) {
        createTextPaintIfNeeded()
        mTextPaint!!.setARGB(mFillStyleA, mFillStyleR, mFillStyleG, mFillStyleB)
        mTextPaint!!.style = Paint.Style.FILL
        scaleX(mTextPaint, text, maxWidth)
        val pt = convertDrawPoint(Point(x, y), text)
        mCanvas.drawText(text, pt.x, pt.y, mTextPaint!!)
    }

    private fun strokeText(text: String, x: Float, y: Float, maxWidth: Float) {
        createTextPaintIfNeeded()
        mTextPaint!!.setARGB(mStrokeStyleA, mStrokeStyleR, mStrokeStyleG, mStrokeStyleB)
        mTextPaint!!.style = Paint.Style.STROKE
        mTextPaint!!.strokeWidth = mLineWidth
        scaleX(mTextPaint, text, maxWidth)
        val pt = convertDrawPoint(Point(x, y), text)
        mCanvas.drawText(text, pt.x, pt.y, mTextPaint!!)
    }

    private fun measureText(text: String): Float {
        createTextPaintIfNeeded()
        return mTextPaint!!.measureText(text)
    }

    private fun measureTextReturnSize(text: String): Size {
        createTextPaintIfNeeded()
        val fm = mTextPaint!!.fontMetrics
        return Size(measureText(text), fm.descent - fm.ascent)
    }

    private fun updateFont(
        fontName: String,
        fontSize: Float,
        bold: Boolean,
        italic: Boolean,
        oblique: Boolean,
        smallCaps: Boolean
    ) {
        mFontName = fontName
        mFontSize = fontSize
        mIsBoldFont = bold
        mIsItalicFont = italic
        mIsObliqueFont = oblique
        mIsSmallCapsFontVariant = smallCaps
        mTextPaint = null // Reset paint to re-create paint object in createTextPaintIfNeeded
    }

    private fun setTextAlign(align: Int) {
        mTextAlign = align
    }

    private fun setTextBaseline(baseline: Int) {
        mTextBaseline = baseline
    }

    private fun setFillStyle(r: Float, g: Float, b: Float, a: Float) {
        mFillStyleR = (r * 255.0f).toInt()
        mFillStyleG = (g * 255.0f).toInt()
        mFillStyleB = (b * 255.0f).toInt()
        mFillStyleA = (a * 255.0f).toInt()
    }

    private fun setStrokeStyle(r: Float, g: Float, b: Float, a: Float) {
        mStrokeStyleR = (r * 255.0f).toInt()
        mStrokeStyleG = (g * 255.0f).toInt()
        mStrokeStyleB = (b * 255.0f).toInt()
        mStrokeStyleA = (a * 255.0f).toInt()
    }

    private fun setLineWidth(lineWidth: Float) {
        mLineWidth = lineWidth
    }

    private fun _fillImageData(
        imageData: ByteArray,
        imageWidth: Float,
        imageHeight: Float,
        offsetX: Float,
        offsetY: Float
    ) {
        val fillSize = (imageWidth * imageHeight).toInt()
        val fillColors = IntArray(fillSize)
        var r: Int
        var g: Int
        var b: Int
        var a: Int
        for (i in 0 until fillSize) {
            r = imageData[4 * i + 0].toInt() and 0xff
            g = imageData[4 * i + 1].toInt() and 0xff
            b = imageData[4 * i + 2].toInt() and 0xff
            a = imageData[4 * i + 3].toInt() and 0xff
            fillColors[i] =
                a and 0xff shl 24 or (r and 0xff shl 16) or (g and 0xff shl 8) or (b and 0xff)
        }
        mBitmap!!.setPixels(
            fillColors,
            0,
            imageWidth.toInt(),
            offsetX.toInt(),
            offsetY.toInt(),
            imageWidth.toInt(),
            imageHeight.toInt()
        )
    }

    private fun convertDrawPoint(point: Point, text: String): Point {
        val ret: Point = Point(point)
        val textSize = measureTextReturnSize(text)
        if (mTextAlign == TEXT_ALIGN_CENTER) {
            ret.x -= textSize.width / 2
        } else if (mTextAlign == TEXT_ALIGN_RIGHT) {
            ret.x -= textSize.width
        }
        if (mTextBaseline == TEXT_BASELINE_TOP) {
            ret.y += textSize.height
        } else if (mTextBaseline == TEXT_BASELINE_MIDDLE) {
            ret.y += textSize.height / 2
        }
        return ret
    }

    private val dataRef: ByteArray?
        get() {
            if (mBitmap != null) {
                val len = mBitmap!!.width * mBitmap!!.height * 4
                val pixels = ByteArray(len)
                val buf = ByteBuffer.wrap(pixels)
                buf.order(ByteOrder.nativeOrder())
                mBitmap!!.copyPixelsToBuffer(buf)
                return pixels
            }
            return null
        }

    companion object {
        private const val TAG = "CanvasContext2D"
        private const val TEXT_ALIGN_LEFT = 0
        private const val TEXT_ALIGN_CENTER = 1
        private const val TEXT_ALIGN_RIGHT = 2
        private const val TEXT_BASELINE_TOP = 0
        private const val TEXT_BASELINE_MIDDLE = 1
        private const val TEXT_BASELINE_BOTTOM = 2
        private var sContext: WeakReference<Context?>? = null
        private const val _sApproximatingOblique = -0.25f //please check paint api documentation
        @JvmStatic
        fun init(context: Context?) {
            sContext = WeakReference(context)
        }

        @JvmStatic
        fun destroy() {
            sContext = null
        }

        private val sTypefaceCache = HashMap<String, Typeface>()

        @JvmStatic
        private fun loadTypeface(familyName: String, urls: String) {
            var url = urls
            if (!sTypefaceCache.containsKey(familyName)) {
                try {
                    var typeface: Typeface? = null
                    if (url.startsWith("/")) {
                        typeface = Typeface.createFromFile(url)
                    } else if (sContext!!.get() != null) {
                        val prefix = "@assets/"
                        if (url.startsWith(prefix)) {
                            url = url.substring(prefix.length)
                        }
                        typeface = Typeface.createFromAsset(
                            CocosAssetsMgr.getAssets(
                                sContext!!.get()!!
                            ), url
                        )
                    }
                    if (typeface != null) {
                        sTypefaceCache[familyName] = typeface
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        @JvmStatic
        private fun clearTypefaceCache() {
            sTypefaceCache.clear()
        }

        @JvmStatic
        private fun newPaint(
            fontName: String,
            fontSize: Int,
            enableBold: Boolean,
            enableItalic: Boolean,
            obliqueFont: Boolean,
            smallCapsFontVariant: Boolean
        ): TextPaint {
            val paint = TextPaint()
            paint.textSize = fontSize.toFloat()
            paint.isAntiAlias = true
            paint.isSubpixelText = true
            var key = fontName
            if (enableBold) {
                key += "-Bold"
                paint.isFakeBoldText = true
            }
            if (enableItalic) {
                key += "-Italic"
            }
            val typeFace: Typeface?
            if (sTypefaceCache.containsKey(key)) {
                typeFace = sTypefaceCache[key]
            } else {
                var style = Typeface.NORMAL
                if (enableBold && enableItalic) {
                    style = Typeface.BOLD_ITALIC
                } else if (enableBold) {
                    style = Typeface.BOLD
                } else if (enableItalic) {
                    style = Typeface.ITALIC
                }
                typeFace = Typeface.create(fontName, style)
            }
            paint.typeface = typeFace
            if (obliqueFont) {
                paint.textSkewX = _sApproximatingOblique
            }
            if (smallCapsFontVariant && Build.VERSION.SDK_INT >= 21) {
                Cocos2dxReflectionHelper.invokeInstanceMethod<Void>(
                    paint,
                    "setFontFeatureSettings",
                    arrayOf(String::class.java),
                    arrayOf("smcp")
                )
            }
            return paint
        }
    }
}